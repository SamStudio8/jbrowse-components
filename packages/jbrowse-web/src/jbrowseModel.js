import {
  ConfigurationSchema,
  readConfObject,
} from '@gmod/jbrowse-core/configuration'
import PluginManager from '@gmod/jbrowse-core/PluginManager'
import RpcManager from '@gmod/jbrowse-core/rpc/RpcManager'
import { getSession } from '@gmod/jbrowse-core/util'
import { openLocation } from '@gmod/jbrowse-core/util/io'
import { flow, getSnapshot, resolveIdentifier, types } from 'mobx-state-tree'
import shortid from 'shortid'
import corePlugins from './corePlugins'
import RenderWorker from './rpc.worker'
import * as rpcFuncs from './rpcMethods'
import AssemblyConfigSchemasFactory from './session/assemblyConfigSchemas'
import sessionModelFactory from './session/sessionModelFactory'

const pluginManager = new PluginManager(corePlugins.map(P => new P()))
pluginManager.configure()

const Session = sessionModelFactory(pluginManager)
const { assemblyConfigSchemas, dispatcher } = AssemblyConfigSchemasFactory(
  pluginManager,
)

const Species = ConfigurationSchema(
  'Species',
  {
    name: {
      type: 'string',
      defaultValue: '',
      description: 'Name of the species',
    },
    assembly: types.union({ dispatcher }, ...assemblyConfigSchemas),
    // track configuration is an array of track config schemas. multiple
    // instances of a track can exist that use the same configuration
    tracks: types.array(pluginManager.pluggableConfigSchemaType('track')),
    connections: types.array(
      pluginManager.pluggableConfigSchemaType('connection'),
    ),
  },
  {
    actions: self => ({
      addTrackConf(typeName, data, connectionName) {
        const type = getSession(self).pluginManager.getTrackType(typeName)
        if (!type) throw new Error(`unknown track type ${typeName}`)
        const schemaType = type.configSchema
        const conf = schemaType.create(Object.assign({ type: typeName }, data))
        if (connectionName) {
          const connectionNames = self.connections.map(connection =>
            readConfObject(connection, 'connectionName'),
          )
          if (!connectionNames.includes(connectionName))
            throw new Error(
              `Cannot add track to non-existent connection: ${connectionName}`,
            )
          self.volatile.get(connectionName).tracks.push(conf)
        } else self.tracks.push(conf)
        return conf
      },
      addConnectionConf(typeName, data) {
        const type = getSession(self).pluginManager.getConnectionType(typeName)
        if (!type) throw new Error(`unknown connection type ${typeName}`)
        const schemaType = type.configSchema
        const conf = schemaType.create(Object.assign({ type: typeName }, data))
        self.connections.push(conf)
        return conf
      },
    }),
  },
)

// poke some things for testing (this stuff will eventually be removed)
window.getSnapshot = getSnapshot
window.resolveIdentifier = resolveIdentifier

const JBrowseWeb = types
  .model('JBrowseWeb', {
    session: types.maybe(Session),
    sessionSnapshots: types.array(types.frozen(Session)),
    species: types.array(Species),
    configuration: ConfigurationSchema('Root', {
      rpc: RpcManager.configSchema,
      // possibly consider this for global config editor
      highResolutionScaling: {
        type: 'number',
        defaultValue: 2,
      },
    }),
  })
  .actions(self => ({
    addSession(sessionConfig) {
      if (sessionConfig.uri || sessionConfig.localPath)
        self.addSessionFromLocation(sessionConfig)
      else
        try {
          let { defaultSession } = sessionConfig
          if (!defaultSession) defaultSession = {}
          if (!defaultSession.menuBars)
            defaultSession.menuBars = [{ type: 'MainMenuBar' }]
          const {
            name = `Unnamed Session ${shortid.generate()}`,
          } = defaultSession

          const data = {
            name,
            ...defaultSession,
            configuration: sessionConfig,
          }
          self.sessions.set(name, data)
          if (!self.session) {
            self.activateSession(name)
          }
        } catch (error) {
          console.error('Failed to add session', error)
        }
    },
    addSessionFromLocation: flow(function* addSessionFromLocation(
      sessionConfigLocation,
    ) {
      try {
        const configSnapshot = JSON.parse(
          yield openLocation(sessionConfigLocation).readFile('utf8'),
        )
        self.addSession(configSnapshot)
      } catch (error) {
        console.error('Failed to fetch config', error)
      }
    }),
    setSession(snapshot) {
      self.session = snapshot
    },
    setEmptySession() {
      self.setSession({
        name: `Unnamed Session ${shortid.generate()}`,
        menuBars: [{ type: 'MainMenuBar' }],
      })
    },
    activateSession(name) {
      const newSessionSnapshot = self.sessionSnapshots.find(
        sessionSnap => sessionSnap.name === name,
      )
      if (!newSessionSnapshot)
        throw new Error(
          `Can't activate session ${name}, it is not in the sessionSnapshots`,
        )
      self.setSession(newSessionSnapshot)
    },

    getRefNameAliases: flow(function* getRefNameAliases(
      assemblyName,
      opts = {},
    ) {
      const refNameAliases = {}
      const assemblyConfig = self.assemblyData.get(assemblyName)
      if (assemblyConfig.refNameAliases) {
        const adapterRefNameAliases = yield self.rpcManager.call(
          assemblyConfig.refNameAliases.adapter.configId,
          'getRefNameAliases',
          {
            sessionId: assemblyName,
            adapterType: assemblyConfig.refNameAliases.adapter.type,
            adapterConfig: assemblyConfig.refNameAliases.adapter,
            signal: opts.signal,
          },
          { timeout: 1000000 },
        )
        adapterRefNameAliases.forEach(alias => {
          refNameAliases[alias.refName] = alias.aliases
        })
      }
      return refNameAliases
    }),

    addRefNameMapForAdapter: flow(function* addRefNameMapForAdapter(
      adapterConf,
      assemblyName,
      opts = {},
    ) {
      const adapterConfigId = readConfObject(adapterConf, 'configId')
      if (!self.refNameMaps.has(adapterConfigId))
        self.refNameMaps.set(adapterConfigId, new Map())
      const refNameMap = self.refNameMaps.get(adapterConfigId)

      const refNameAliases = yield self.getRefNameAliases(assemblyName, opts)

      const refNames = yield self.rpcManager.call(
        readConfObject(adapterConf, 'configId'),
        'getRefNames',
        {
          sessionId: assemblyName,
          adapterType: readConfObject(adapterConf, 'type'),
          adapterConfig: adapterConf,
          signal: opts.signal,
        },
        { timeout: 1000000 },
      )
      refNames.forEach(refName => {
        if (refNameAliases[refName])
          refNameAliases[refName].forEach(refNameAlias => {
            refNameMap.set(refNameAlias, refName)
          })
        else
          Object.keys(refNameAliases).forEach(configRefName => {
            if (refNameAliases[configRefName].includes(refName)) {
              refNameMap.set(configRefName, refName)
              refNameAliases[configRefName].forEach(refNameAlias => {
                if (refNameAlias !== refName)
                  refNameMap.set(refNameAlias, refName)
              })
            }
          })
      })
      return refNameMap
    }),

    getRefNameMapForAdapter: flow(function* getRefNameMapForAdapter(
      adapterConf,
      assemblyName,
      opts = {},
    ) {
      const configId = readConfObject(adapterConf, 'configId')
      if (!self.refNameMaps.has(configId)) {
        return yield self.addRefNameMapForAdapter(
          adapterConf,
          assemblyName,
          opts,
        )
      }
      return self.refNameMaps.get(configId)
    }),
  }))
  .views(self => ({
    get sessionNames() {
      return self.sessionSnapshots.map(sessionSnap => sessionSnap.name)
    },

    get assemblyData() {
      const assemblyData = new Map()
      for (const speciesConfig of self.species) {
        const assemblyConfig = speciesConfig.assembly
        const assemblyName = readConfObject(assemblyConfig, 'name')
        const assemblyInfo = {}
        if (assemblyConfig.sequence)
          assemblyInfo.sequence = assemblyConfig.sequence
        const refNameAliasesConf = readConfObject(
          assemblyConfig,
          'refNameAliases',
        )
        if (refNameAliasesConf) assemblyInfo.refNameAliases = refNameAliasesConf
        const aliases = readConfObject(assemblyConfig, 'aliases')
        assemblyInfo.aliases = aliases
        assemblyData.set(assemblyName, assemblyInfo)
        aliases.forEach((assemblyAlias, idx) => {
          const newAliases = [
            ...aliases.slice(0, idx),
            ...aliases.slice(idx + 1),
            assemblyName,
          ]
          assemblyData.set(assemblyAlias, {
            ...assemblyInfo,
            aliases: newAliases,
          })
        })
      }
      for (const assemblyName of self.session.connections.keys()) {
        const connectionConfs = self.session.connections.get(assemblyName)
        // eslint-disable-next-line no-loop-func
        connectionConfs.forEach(connectionConf => {
          if (!assemblyData.has(assemblyName)) {
            const assemblyInfo = {}
            assemblyInfo.sequence = connectionConf.sequence
            assemblyInfo.refNameAliases = connectionConf.refNameAliases
            assemblyData.set(assemblyName, assemblyInfo)
          } else {
            if (
              !assemblyData.get(assemblyName).refNameAliases &&
              connectionConf.refNameAliases
            ) {
              assemblyData.get(assemblyName).refNameAliases = readConfObject(
                connectionConf.refNameAliases,
              )
              assemblyData.get(assemblyName).aliases.forEach(alias => {
                assemblyData.get(alias).refNameAliases = readConfObject(
                  connectionConf.refNameAliases,
                )
              })
            }
            if (
              (!assemblyData.get(assemblyName).sequence &&
                connectionConf.sequence) ||
              connectionConf.defaultSequence
            ) {
              assemblyData.get(assemblyName).sequence = connectionConf.sequence
              assemblyData.get(assemblyName).aliases.forEach(alias => {
                assemblyData.get(alias).sequence = connectionConf.sequence
              })
            }
          }
        })
      }
      return assemblyData
    },
  }))
  .volatile(self => ({
    rpcManager: new RpcManager(
      pluginManager,
      self.configuration.rpc,
      {
        WebWorkerRpcDriver: { WorkerClass: RenderWorker },
        MainThreadRpcDriver: { rpcFuncs },
      },
      self.getRefNameMapForAdapter,
    ),
    refNameMaps: new Map(),
  }))

export function createTestSession(snapshot = {}, root = false) {
  const jbrowseState = JBrowseWeb.create({
    configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
  })
  jbrowseState.addSession(snapshot)
  if (root) return jbrowseState
  return jbrowseState.activeSession
}

export default JBrowseWeb
