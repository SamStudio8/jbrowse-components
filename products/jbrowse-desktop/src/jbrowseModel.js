import {
  ConfigurationSchema,
  readConfObject,
} from '@jbrowse/core/configuration'
import RpcManager from '@jbrowse/core/rpc/RpcManager'
import {
  getParent,
  getSnapshot,
  resolveIdentifier,
  types,
  walk,
  getType,
  getMembers,
  isModelType,
  isReferenceType,
} from 'mobx-state-tree'

// poke some things for testing (this stuff will eventually be removed)
window.getSnapshot = getSnapshot
window.resolveIdentifier = resolveIdentifier

export default function JBrowseDesktop(
  pluginManager,
  Session,
  assemblyConfigSchemasType,
) {
  return types
    .model('JBrowseDesktop', {
      configuration: ConfigurationSchema('Root', {
        rpc: RpcManager.configSchema,
        // possibly consider this for global config editor
        highResolutionScaling: {
          type: 'number',
          defaultValue: 2,
        },
        useUrlSession: {
          type: 'boolean',
          defaultValue: true,
        },
        useLocalStorage: {
          type: 'boolean',
          defaultValue: false,
        },
      }),
      plugins: types.frozen(),
      assemblies: types.array(assemblyConfigSchemasType),
      // track configuration is an array of track config schemas. multiple
      // instances of a track can exist that use the same configuration
      tracks: types.array(pluginManager.pluggableConfigSchemaType('track')),
      connections: types.array(
        pluginManager.pluggableConfigSchemaType('connection'),
      ),
      defaultSession: types.optional(types.frozen(Session), {
        name: `New Session`,
      }),
    })
    .actions(self => ({
      afterCreate() {
        const seen = []
        self.assemblyNames.forEach(assemblyName => {
          if (!assemblyName) {
            throw new Error('Encountered an assembly with no "name" defined')
          }
          if (seen.includes(assemblyName)) {
            throw new Error(
              `Found two assemblies with the same name: ${assemblyName}`,
            )
          } else {
            seen.push(assemblyName)
          }
        })
      },
      addSavedSession(sessionSnapshot) {
        const length = self.savedSessions.push(sessionSnapshot)
        return self.savedSessions[length - 1]
      },
      removeSavedSession(sessionSnapshot) {
        self.savedSessions.remove(sessionSnapshot)
      },
      replaceSavedSession(oldName, snapshot) {
        const savedSessionIndex = self.savedSessions.findIndex(
          savedSession => savedSession.name === oldName,
        )
        self.savedSessions[savedSessionIndex] = snapshot
      },
      addAssemblyConf(assemblyConf) {
        const { name } = assemblyConf
        if (!name) throw new Error('Can\'t add assembly with no "name"')
        if (self.assemblyNames.includes(name))
          throw new Error(
            `Can't add assembly with name "${name}", an assembly with that name already exists`,
          )
        const length = self.assemblies.push(assemblyConf)
        return self.assemblies[length - 1]
      },
      addTrackConf(trackConf) {
        const { type } = trackConf
        if (!type) throw new Error(`unknown track type ${type}`)
        const length = self.tracks.push(trackConf)
        return self.tracks[length - 1]
      },
      addConnectionConf(connectionConf) {
        const { type } = connectionConf
        if (!type) throw new Error(`unknown connection type ${type}`)
        const length = self.connections.push(connectionConf)
        return self.connections[length - 1]
      },

      hasWidget(widget) {
        return getParent(self).session.activeWidgets.has(widget.id)
      },
      removeReferring(referring, track, callbacks, dereferenceTypeCount) {
        referring.forEach(({ node }) => {
          let dereferenced = false
          try {
            // If a view is referring to the track config, remove the track
            // from the view
            const type = 'open track(s)'
            const view = getContainingView(node)
            callbacks.push(() => view.hideTrack(track.trackId))
            dereferenced = true
            if (!dereferenceTypeCount[type]) {
              dereferenceTypeCount[type] = 0
            }
            dereferenceTypeCount[type] += 1
          } catch (err1) {
            // ignore
          }
          if (this.hasWidget(node)) {
            // If a configuration editor widget has the track config
            // open, close the widget
            const type = 'configuration editor widget(s)'
            callbacks.push(() => getParent(self).session.hideWidget(node))
            dereferenced = true
            if (!dereferenceTypeCount[type]) dereferenceTypeCount[type] = 0
            dereferenceTypeCount[type] += 1
          }
          if (!dereferenced)
            throw new Error(
              `Error when closing this connection, the following node is still referring to a track configuration: ${JSON.stringify(
                getSnapshot(node),
              )}`,
            )
        })
      },
      getReferring(object) {
        const refs = []
        walk(getParent(self), node => {
          if (isModelType(getType(node))) {
            const members = getMembers(node)
            Object.entries(members.properties).forEach(([key, value]) => {
              // @ts-ignore
              if (isReferenceType(value) && node[key] === object) {
                refs.push({ node, key })
              }
            })
          }
        })
        return refs
      },
      deleteTrackConf(trackConf) {
        const { trackId } = trackConf
        const idx = self.tracks.findIndex(t => t.trackId === trackId)
        if (idx === -1) {
          return undefined
        }
        const callbacksToDereferenceTrack = []
        const dereferenceTypeCount = {}
        const referring = self.getReferring(trackConf)
        this.removeReferring(
          referring,
          trackConf,
          callbacksToDereferenceTrack,
          dereferenceTypeCount,
        )
        callbacksToDereferenceTrack.forEach(cb => cb())
        return self.tracks.splice(idx, 1)
      },
    }))
    .views(self => ({
      get savedSessionNames() {
        return getParent(self).savedSessionNames
      },
      get assemblyNames() {
        return self.assemblies.map(assembly => readConfObject(assembly, 'name'))
      },
      get rpcManager() {
        return getParent(self).rpcManager
      },
    }))
}
