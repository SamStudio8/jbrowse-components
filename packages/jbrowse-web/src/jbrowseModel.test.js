import { getSnapshot } from 'mobx-state-tree'
import jbrowseModel from './jbrowseModel'

describe('JBrowse MST model', () => {
  it('creates with defaults', () => {
    const jbrowse = jbrowseModel.create()
    expect(jbrowse.session).toBeUndefined()
    jbrowse.setEmptySession()
    expect(jbrowse.session).toBeTruthy()
    expect(jbrowse.sessionSnapshots.length).toBe(0)
    expect(jbrowse.species.length).toBe(0)
    expect(getSnapshot(jbrowse.configuration)).toMatchSnapshot({
      configId: expect.any(String),
      rpc: { configId: expect.any(String) },
    })
  })

  it('creates with a minimal session', () => {
    const jbrowse = jbrowseModel.create({ session: { name: 'testSession' } })
    expect(jbrowse.session).toBeTruthy()
  })

  it('activates a session snapshot', () => {
    const jbrowse = jbrowseModel.create({
      sessionSnapshots: [{ name: 'testSession' }],
    })
    expect(jbrowse.session).toBeUndefined()
    jbrowse.setSession(jbrowse.sessionSnapshots[0])
    expect(jbrowse.session).toBeTruthy()
  })

  it('adds track and connection configs to a species', () => {
    const jbrowse = jbrowseModel.create({
      species: [
        {
          name: 'species1',
          assembly: { name: 'assembly1', aliases: ['assemblyA'] },
        },
      ],
    })
    expect(jbrowse.species.length).toBe(1)
    expect(getSnapshot(jbrowse.species[0])).toMatchSnapshot({
      configId: expect.any(String),
      assembly: {
        configId: expect.any(String),
      },
    })
    const newTrackConf = jbrowse.species[0].addTrackConf({ type: 'BasicTrack' })
    expect(getSnapshot(newTrackConf)).toMatchSnapshot({
      configId: expect.any(String),
      adapter: {
        configId: expect.any(String),
        index: { configId: expect.any(String) },
      },
      renderer: { configId: expect.any(String) },
    })
    expect(jbrowse.species[0].tracks.length).toBe(1)
    const newConnectionConf = jbrowse.species[0].addConnectionConf({
      type: 'JBrowse1Connection',
    })
    expect(getSnapshot(newConnectionConf)).toMatchSnapshot({
      configId: expect.any(String),
    })
    expect(jbrowse.species[0].connections.length).toBe(1)
  })

  it('adds a session snapshot', () => {
    const jbrowse = jbrowseModel.create()
    jbrowse.addSessionSnapshot({ name: 'testSession' })
    expect(jbrowse.sessionSnapshots.length).toBe(1)
  })

  it('throws if session is invalid', () => {
    expect(() => jbrowseModel.create({ session: {} })).toThrow()
  })

  it('throws if session snapshot is invalid', () => {
    expect(() => jbrowseModel.create({ sessionSnapshots: [{}] })).toThrow()
  })
})
