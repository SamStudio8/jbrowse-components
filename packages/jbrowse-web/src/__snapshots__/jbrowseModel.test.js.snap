// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JBrowse MST model adds track and connection configs to a species 1`] = `
Object {
  "assembly": Object {
    "aliases": Array [
      "assemblyA",
    ],
    "configId": Any<String>,
    "name": "assembly1",
  },
  "configId": Any<String>,
  "name": "species1",
}
`;

exports[`JBrowse MST model adds track and connection configs to a species 2`] = `
Object {
  "adapter": Object {
    "configId": Any<String>,
    "index": Object {
      "configId": Any<String>,
    },
    "type": "BamAdapter",
  },
  "configId": Any<String>,
  "renderer": Object {
    "configId": Any<String>,
    "type": "PileupRenderer",
  },
  "type": "BasicTrack",
}
`;

exports[`JBrowse MST model adds track and connection configs to a species 3`] = `
Object {
  "configId": Any<String>,
  "type": "JBrowse1Connection",
}
`;

exports[`JBrowse MST model creates with defaults 1`] = `
Object {
  "configId": Any<String>,
  "rpc": Object {
    "configId": Any<String>,
  },
}
`;
